core/dbt/materializations/incremental/microbatch.py

```diff
@@ -35,10 +36,13 @@ def __init__(
             event_time_start.replace(tzinfo=pytz.UTC) if event_time_start else None
         )
         self.event_time_end = event_time_end.replace(tzinfo=pytz.UTC) if event_time_end else None
+        self.batch_current_time = (
+            batch_current_time if batch_current_time is not None else datetime.now(pytz.UTC)
+        )
 
     def build_end_time(self):
         """Defaults the end_time to the current time in UTC unless a non `None` event_time_end was provided"""
-        return self.event_time_end or datetime.now(tz=pytz.utc)
+        return self.event_time_end or self.batch_current_time
```

Comment:
Under which situations will we return `self.batch_current_time` here?
Probably worth adding to the description.

Reply:
From the function perspective, we return `self.batch_current_time` when `self.event_time_end` is `None`. From a system perspective this currently happens when the dbt runner is not invoked with an `--event-time-end <my-desired-end-value>`. I don't think the system context is relevant to this function though. If anything I think we should separately push that context to the instantiation of the class in `run.py` and drop this `or`-ing from the class itself.

I do have a planned refactor of this class for this week, this context shifting would be more appropriate as part of that.